python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1303 @@
+                xcorrs[det_object].append(self._crossCorrelate(_scaledMaskedIm1[det_object],
+                                                               _scaledMaskedIm2[det_object]))
+                means[det_object].append([_means1[det_object], _means2[det_object]])
+
+            # this is awkward now that code is refactored
+            # not sure how best to make this work without passing everything around
+            # if self.debug.enabled:
+            #     rawMeans = [afwMath.makeStatistics(im.getMaskedImage(), afwMath.MEANCLIP).getValue() \
+            #                 for im in [exp1, exp2]]
+            #     title = "Visits %s; %s, Detectors %s <I> = ",
+            #             "%.3f (%s) Var = %.4f" % (self._getNameOfSet([v1]),
+            #                                       self._getNameOfSet([v2]),
+            #                                       self._getNameOfSet([detNum]),
+            #                                       _means1+rawMeans[1],
+            #                                       im1.getFilter().getName(),
+            #                                       float(xcorr[0, 0]) / (_means1+rawMeans[1]))
+            #
+            #     detObjId = str(detNum)
+            #     if level=='AMP': detObjId += ampName
+            #     fileName = (os.path.join(self.debug.debugPlotPath, '_'.join(['xcorr_visit', str(v1),
+            #                                                                  str(v2), 'detector',
+            #                                                                  str(detNum)])))
+            #     fileName += self.debug.plotType
+            #     self._plotXcorr(xcorr.copy(), (xcorrMeans[0]+means[1]),
+            #                     title=title, saveToFileName=fileName)
+
+        # generate the kernel(s)
+        self.log.info('Generating kernel(s) for %s' % detNum)
+        for det_object in xcorrs.keys():  # looping over either detectors or amps
+            if self.config.level == 'DETECTOR':
+                objId = 'detector %s' % det_object
+            elif self.config.level == 'AMP':
+                objId = 'detector %s AMP %s' % (detNum, det_object)
+            kernels[det_object] = self.generateKernel(xcorrs[det_object], means[det_object], objId)
+        # kernels['level'] = self.config.level  # this might be useful, but might also mess things up
+        dataRef.put(kernels)
+
+        self.log.info('Finished generating kernel(s) for %s' % detNum)
+        return pipeBase.Struct(exitStatus=0)
+
+    def _prepareImage(self, exp, gains, level):
+        """Prepare images for cross-correlation calculation.
+
+        Each amp has borders applied, is rescaled by the gain, and has the sigma-clipped mean subtracted.
+        If the level is 'DETECTOR' then this is done for the whole image in-place so that it can be
+        cross-correlated.
+        If the level is 'AMP' then this is done per-amplifier, and each amp-image returned.
```

Comment:
Please consider rewriting in active voice (the usual Python standard for doc strings) e.g. "For each amp apply borders (what does that mean?) rescale by gain, and subtract sigma-clipped mean"...